File: timetable-backend/app/schemas/timetable.py
"""Pydantic schemas for timetable API."""
from pydantic import BaseModel
from typing import Optional
from datetime import datetime


class FacultySchema(BaseModel):
    """Faculty schema."""
    name: str
    email: str
    department: str
    max_hours_per_week: Optional[int] = 20


class CourseSchema(BaseModel):
    """Course schema."""
    code: str
    name: str
    credits: float
    hours_per_week: int
    semester: int


class SectionSchema(BaseModel):
    """Section schema."""
    name: str
    course_id: int
    capacity: int
    faculty_id: Optional[int] = None


class RoomSchema(BaseModel):
    """Room schema."""
    code: str
    name: str
    capacity: int
    floor: Optional[int] = None


class TimeSlotSchema(BaseModel):
    """TimeSlot schema."""
    day: str
    start_time: str
    end_time: str
    duration_minutes: int


class TimetableSchema(BaseModel):
    """Timetable entry schema."""
    section_id: int
    faculty_id: int
    room_id: int
    timeslot_id: int
    status: Optional[str] = "scheduled"

    class Config:
        from_attributes = True


class TimetableResponseSchema(TimetableSchema):
    """Timetable response schema."""
    id: int
    created_at: datetime
    updated_at: datetime

File: timetable-backend-mern/models/Timetable.js
const mongoose = require('mongoose');

const timetableSchema = new mongoose.Schema({
    sectionId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Section',
        required: true,
    },
    courseId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Course',
        required: true,
    },
    facultyId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Faculty',
        required: true,
    },
    roomId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Room',
        required: true,
    },
    timeslotId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'TimeSlot',
        required: true,
    },
}, {
    timestamps: true,
});

module.exports = mongoose.model('Timetable', timetableSchema);
